yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -51,6 +51,38 @@ private[spark] abstract class YarnSchedulerBackend(
 
   private implicit val askTimeout = RpcUtils.askRpcTimeout(sc.conf)
 
+  /** Application ID. Must be set by a subclass before starting the service */
+  private var appId: ApplicationId = null
+
+  /** Attempt ID. This is unset for client-side schedulers */
+  private var attemptId: Option[ApplicationAttemptId] = None
+
+  /** Scheduler extension services */
+  private val services: SchedulerExtensionServices = new SchedulerExtensionServices()
+
+  /**
+    * Bind to YARN. This *must* be done before calling [[start()]].
+    *
+    * @param appId YARN application ID
+    * @param attemptId Optional YARN attempt ID
+    */
+  protected def bindToYarn(appId: ApplicationId, attemptId: Option[ApplicationAttemptId]): Unit = {
+    this.appId = appId
+    this.attemptId = attemptId
+  }
+
+  override def start() {
+    require(appId != null, "application ID unset")
+    val binding = SchedulerExtensionServiceBinding(sc, appId, attemptId)
+    services.start(binding)
+    super.start()
+  }
+
+  override def stop(): Unit = {
+    super.stop()
```

Comment:
Good question about what to do with exceptions. I think the best approach is to have `stop()` be called from a method that catches and logs exceptions. That way you kill two problems: you'll not replace any exceptions caused by `super.stop()`, and you'll make sure `stop()` is called for all existing services (which the current code doesn't do when exceptions are thrown).